src/Microsoft.AspNet.Mvc.Core/SessionStateTempDataProvider.cs

```diff
@@ -37,9 +68,62 @@ public class SessionStateTempDataProvider : ITempDataProvider
                 using (var memoryStream = new MemoryStream(value))
                 using (var writer = new BsonReader(memoryStream))
                 {
-                    tempDataDictionary = jsonSerializer.Deserialize<Dictionary<string, object>>(writer);
+                    tempDataDictionary = _jsonSerializer.Deserialize<Dictionary<string, object>>(writer);
+                }
+
+                var convertedDictionary = new Dictionary<string, object>(tempDataDictionary, StringComparer.OrdinalIgnoreCase);
+                foreach (var item in tempDataDictionary)
+                {
+                    var jArrayValue = item.Value as JArray;
+                    if (jArrayValue != null && jArrayValue.Count > 0)
+                    {
+                        Type returnType;
+                        if (_tokenTypeLookup.TryGetValue(jArrayValue[0].Type, out returnType))
```

Comment:
Could you refactor this out in to a variable:
`var arrayType = jArrayValue[0].Type;`